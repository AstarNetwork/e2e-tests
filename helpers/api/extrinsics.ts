import { AccountInfo } from '@polkadot/types/interfaces'
import { ApiPromise } from '@polkadot/api'
import { SubmittableExtrinsic } from '@polkadot/api-base/types'

export const balance = async (api: ApiPromise, address: string) => {
  const account = await api.query.system.account<AccountInfo>(address)
  return account.data.toJSON()
}

export const xTokensForRelayChain = (api: ApiPromise, amount: string, address: Uint8Array) => {
  return api.tx.xTokens.transfer(
    {
      Token: 'KSM',
    },
    amount,
    {
      V1: {
        parents: 1,
        interior: {
          X1: {
            AccountId32: {
              network: 'Any',
              id: address,
            },
          },
        },
      },
    },
    'Unlimited',
  )
}

export const xTokensForParaChain = (
  api: ApiPromise,
  token: object,
  parachainId: string,
  amount: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transfer(
    token,
    amount,
    {
      V1: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                network: 'Any',
                id: address,
              },
            },
          ],
        },
      },
    },
    'Unlimited',
  )
}

export const xTokens = (
  api: ApiPromise,
  isRelayChain: boolean,
  parachainId: string,
  token: object,
  amount: string | bigint,
  address: Uint8Array,
) => {
  const multiLocation = isRelayChain
    ? { X1: { AccountId32: { network: 'Any', id: address } } }
    : {
        X2: [
          {
            Parachain: parachainId,
          },
          {
            AccountId32: {
              network: 'Any',
              id: address,
            },
          },
        ],
      }

  return api.tx.xTokens.transfer(
    token,
    amount,
    {
      V1: {
        parents: 1,
        interior: multiLocation,
      },
    },
    'Unlimited',
  )
}

export const xTokensV3 = (
  api: ApiPromise,
  isRelayChain: boolean,
  parachainId: string,
  token: object,
  amount: string | bigint,
  address: Uint8Array,
) => {
  const multiLocation = isRelayChain
    ? {
        X1: {
          AccountId32: {
            id: address,
          },
        },
      }
    : {
        X2: [
          {
            Parachain: parachainId,
          },
          {
            AccountId32: {
              id: address,
            },
          },
        ],
      }

  return api.tx.xTokens.transfer(
    token,
    amount,
    {
      V3: {
        parents: 1,
        interior: multiLocation,
      },
    } as any,
    'Unlimited',
  )
}

export const relayChainV3limitedReserveTransferAssets = (
  api: ApiPromise,
  parachainId: string,
  amount: string,
  address: Uint8Array,
) => {
  return api.tx.xcmPallet.limitedReserveTransferAssets(
    {
      V3: {
        parents: 0,
        interior: {
          X1: { Parachain: parachainId },
        },
      },
    },
    {
      V3: {
        parents: 0,
        interior: {
          X1: {
            AccountId32: {
              id: address,
            },
          },
        },
      },
    },
    {
      V3: [
        {
          id: { Concrete: { parents: 0, interior: 'Here' } },
          fun: { Fungible: amount },
        },
      ],
    },
    0,
    'Unlimited',
  )
}

export const xTokensTransferMulticurrencies = (
  api: ApiPromise,
  foreignAssetId: string,
  amount: string,
  parachainId: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transferMulticurrencies(
    [
      [
        {
          ForeignAsset: foreignAssetId,
        },
        amount,
      ],
      [
        {
          Token: 'KSM',
        },
        '16000000000',
      ],
    ],
    '1',
    {
      V1: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                network: 'Any',
                id: address,
              },
            },
          ],
        },
      },
    },
    'Unlimited',
  )
}

export const xTokensTransferTransferMultiasset = (
  api: ApiPromise,
  interior: any[],
  amount: string,
  parachainId: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transferMultiasset(
    {
      V1: {
        fun: {
          Fungible: amount,
        },
        id: {
          Concrete: {
            parents: 1,
            interior: {
              X3: interior,
            },
          },
        },
      },
    },
    {
      V1: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                network: 'Any',
                id: address,
              },
            },
          ],
        },
      },
    },
    'Unlimited',
  )
}

export const xTokensTransferTransferMultiassetV3 = (
  api: ApiPromise,
  interior: any[],
  amount: string,
  parachainId: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transferMultiasset(
    {
      V3: {
        fun: {
          Fungible: amount,
        },
        id: {
          Concrete: {
            parents: 1,
            interior: {
              X3: interior,
            },
          },
        },
      },
    },
    {
      V3: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                id: address,
              },
            },
          ],
        },
      },
    } as any,
    'Unlimited',
  )
}

export const xTokensTransferMulticurrenciesV3 = (
  api: ApiPromise,
  foreignAssetId: string,
  amount: string,
  parachainId: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transferMulticurrencies(
    [
      [
        {
          ForeignAsset: foreignAssetId,
        },
        amount,
      ],
      [
        {
          Token: 'KSM',
        },
        '16000000000',
      ],
    ],
    '1',
    {
      V3: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                id: address,
              },
            },
          ],
        },
      },
    } as any,
    'Unlimited',
  )
}

export const sudo = (api: ApiPromise, call: SubmittableExtrinsic<'promise'>) => {
  return api.tx.sudo.sudo(call)
}
