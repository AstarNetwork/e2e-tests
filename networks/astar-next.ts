import { Config } from './types'

import astarConfig, { Vars } from './astar'

export default {
  ...astarConfig,
  polkadot: {
    ...astarConfig.polkadot,
    name: 'astarNext' as const,
  },
  kusama: {
    ...astarConfig.kusama,
    name: 'shidenNext' as const,
  },
  config: (opt) => ({
    ...astarConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/astar_runtime.wasm',
        kusama: './wasm/shiden_runtime.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
