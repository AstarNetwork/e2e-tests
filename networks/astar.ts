import { Config } from './types'

export type Vars = {
  relayToken: string
}

export default {
  polkadot: {
    name: 'astar' as const,
    endpoint: 'wss://astar.api.onfinality.io/public-ws',
    relayToken: '340282366920938463463374607431768211455',
  },
  kusama: {
    name: 'shiden' as const,
    endpoint: 'wss://shiden.api.onfinality.io/public-ws',
    relayToken: '340282366920938463463374607431768211455',
  },
  config: ({ alice, relayToken }) => ({
    storages: {
      System: {
        Account: [[[alice.address], { data: { free: 1000n * 10n **18n } }]],
      },
      Assets: {
        account: [[[relayToken, alice.address], { balance: 10n ** 18n }]],
      },
      Sudo: {
        key: alice.address,
      },
      PolkadotXcm: {
        // avoid sending xcm version change notifications to makes things faster
        $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
      },
    },
  }),
} satisfies Config<Vars>

export const astar = {
  paraId: 2006,
  paraAccount: '13YMK2eZzuFY1WZGagpYtTgbWBWGdoUD2CtrPj1mQPjY8Ldc',
  dot: 340282366920938463463374607431768211455n,
  astr: { Concrete: { parents: 0, interior: 'Here' } },
  aca: 18446744073709551616n,
  usdt: 4294969280n,
  dot_loc : { Concrete: { parents: 1, interior: 'Here' } },
  usdt_loc: { Concrete: { parents: 1, interior: { X3: [ { Parachain: 1000 }, { PalletInstance: 50 }, { GeneralIndex: 1984 }] } } },
} as const

export const shiden = {
  paraId: 2007,
  paraAccount: 'F7fq1jNy74AqkJ1DP4KqSrWtnTGtXfNVoDwFhTvvPxUvJaq',
  ksm: 340282366920938463463374607431768211455n,
  sdn: { Concrete: { parents: 0, interior: 'Here' } },
  kar: 18446744073709551618n,
  usdt: 4294969280n,
  ausd: 18446744073709551616n,
} as const
