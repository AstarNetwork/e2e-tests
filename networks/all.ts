import { Config } from './types'

import astar from './astar'
import astarNext from './astar-next'
import polkadot from './polkadot'
import statemint from './statemint'

const all = {
  polkadot,
  statemint,
  astar,
  astarNext,
} satisfies Record<string, Config>

export default all
