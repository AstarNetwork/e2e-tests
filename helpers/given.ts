import { test } from 'vitest'
import _ from 'lodash'

import { Network, NetworkNames, createContext, createNetworks } from '../networks'

class NetworkManager {
  keyring = createContext()

  async buildNetworks(names: NetworkNames[]) {
    const networkOptions = _.zipObject(_.zip(names) as any) as any as Record<NetworkNames, undefined>
    return createNetworks(networkOptions, this.keyring)
  }
}

export const given = <T extends NetworkNames>(...names: NetworkNames[]) =>
  test.extend<{
    networks: Record<T, Network>
    keyring: ReturnType<typeof createContext>
  }>({
    networks: async ({ task }, use) => {
      if (task.retry) {
        throw new Error('given cannot run with retry')
      }
      if (task.concurrent) {
        throw new Error('given cannot run in concurrent mode')
      }
      const manager = new NetworkManager()
      _.assign(task.suite, { __networkManager__: manager })
      await use(await manager.buildNetworks(names))
    },
    keyring: async ({ task }, use) => {
      // network manager is assigned above
      const manager: NetworkManager = _.get(task.suite, '__networkManager__')!
      await use(manager.keyring)
    },
  })
