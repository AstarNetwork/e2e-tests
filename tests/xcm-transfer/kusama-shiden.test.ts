import { checkSystemEvents, given } from '../../helpers'
import { describe, expect } from 'vitest'
import { kusama as kusamaUtil } from '../../networks/polkadot'
import { query, tx } from '../../helpers/api'
import { shiden as shidenUtil } from '../../networks/astar'
import { KusamaMigationStep, AssetHubMigrationStep } from './asset_hub_migration_config.ts'

describe('Kusama & Shiden', () => {
  given('kusama', 'shiden')(
    '001: Kusama transfer KSM to Shiden',
    async ({ networks: { kusama, shiden }, keyring: { alice, bob } }) => {
      await tx.xcmPallet
        .limitedReserveTransferAssetsV3(kusamaUtil.ksm, 1e12, tx.xcmPallet.parachainV3(
            0,
            shidenUtil.paraId,
          ))(kusama, bob.addressRaw)
        .signAndSend(alice)

      await kusama.chain.newBlock()

      if (KusamaMigationStep === AssetHubMigrationStep.NotStrated) {
        await checkSystemEvents(kusama, 'xcmPallet').toMatchSnapshot('001: kusama event')

        await shiden.chain.newBlock()

        const bobBalance = await query.assets(shidenUtil.ksm)(shiden, bob.address)
        expect(bobBalance.unwrap().balance.toNumber()).closeTo(
          1e12,
          1e6, // some fee
          'Expected amount was not received',
        )

        await checkSystemEvents(shiden, 'parachainSystem', 'dmpQueue', 'messageQueue').toMatchSnapshot(
          '003: shiden event',
        )
      } else if (KusamaMigationStep === AssetHubMigrationStep.Ongoing) {
        // Do not check snapshot

        // Tokens should not have been received, and balance is none
        const bobBalance = await query.assets(shidenUtil.ksm)(shiden, bob.address)
        expect(bobBalance.isNone, 'Expected no balance').toBe(true)
      }
    },
  )

  given('kusama', 'shiden')(
    '002: Shiden transfer DOT to Kusama',
    async ({ networks: { shiden, kusama }, keyring: { alice, bob } }) => {
      await tx.xtokens
        .transfer(shidenUtil.ksm, 1e12, tx.xtokens.relaychainV3)(shiden, bob.addressRaw)
        .signAndSend(alice)

      await shiden.chain.newBlock()

      if (KusamaMigationStep === AssetHubMigrationStep.NotStrated) {
        await checkSystemEvents(shiden, 'parachainSystem', 'dmpQueue', 'messageQueue').toMatchSnapshot(
          '001: shiden event',
        )

        await kusama.chain.newBlock()

        const bobBalance = await kusama.api.query.system.account(bob.address)
        expect(bobBalance.data.free.toNumber()).closeTo(
          0,
          1e8, // some fee
          'Expected amount was not received',
        )

        await checkSystemEvents(kusama, 'messageQueue').toMatchSnapshot('002: kusama event')
      } else if (KusamaMigationStep === AssetHubMigrationStep.Ongoing) {
        // Do not check snapshot

        // Tokens should not have been received
        const bobBalance = await kusama.api.query.system.account(bob.address)
        expect(bobBalance.data.free.toNumber()).closeTo(
          0,
          0, // some fee
          'Should be 0',
        )
      }
    },
  )
})
