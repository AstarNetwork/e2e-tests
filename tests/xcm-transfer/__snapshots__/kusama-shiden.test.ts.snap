// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Kusama & Shiden > 001: Kusama transfer KSM to Shiden > 001: kusama event 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "used": {
            "proofSize": "(rounded 6200)",
            "refTime": "(rounded 360000000)",
          },
        },
      },
    },
    "method": "Attempted",
    "section": "xcmPallet",
  },
  {
    "data": {
      "fees": [
        {
          "fun": {
            "Fungible": "(rounded 1300000000)",
          },
          "id": {
            "interior": "Here",
            "parents": 0,
          },
        },
      ],
      "paying": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Kusama",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "FeesPaid",
    "section": "xcmPallet",
  },
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": "(rounded 2000)",
            },
          ],
        },
        "parents": 0,
      },
      "message": [
        {
          "ReserveAssetDeposited": [
            {
              "fun": {
                "Fungible": 1000000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 1000000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "DepositAsset": {
            "assets": {
              "Wild": {
                "AllCounted": 1,
              },
            },
            "beneficiary": {
              "interior": {
                "X1": [
                  {
                    "AccountId32": {
                      "id": "(hash)",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Kusama",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "xcmPallet",
  },
]
`;

exports[`Kusama & Shiden > 001: Kusama transfer KSM to Shiden > 003: shiden event 1`] = `
[
  {
    "data": {
      "count": 1,
    },
    "method": "DownwardMessagesReceived",
    "section": "parachainSystem",
  },
  {
    "data": {
      "dmqHead": "(hash)",
      "weightUsed": {
        "proofSize": "(rounded 8000)",
        "refTime": "(rounded 530000000)",
      },
    },
    "method": "DownwardMessagesProcessed",
    "section": "parachainSystem",
  },
  {
    "data": {
      "id": "(hash)",
      "origin": "Parent",
      "success": true,
      "weightUsed": {
        "proofSize": "(rounded 9300)",
        "refTime": "(rounded 900000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`Kusama & Shiden > 002: Shiden transfer DOT to Kusama > 001: shiden event 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "UpwardMessageSent",
    "section": "parachainSystem",
  },
]
`;

exports[`Kusama & Shiden > 002: Shiden transfer DOT to Kusama > 002: kusama event 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Ump": {
          "Para": "(rounded 2000)",
        },
      },
      "success": true,
      "weightUsed": {
        "proofSize": "(rounded 7200)",
        "refTime": "(rounded 360000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;
